packages/opentelemetry-plugin-fetch/src/fetch.ts

```diff
@@ -0,0 +1,350 @@
+/*!
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as shimmer from 'shimmer';
+import * as api from '@opentelemetry/api';
+import * as core from '@opentelemetry/core';
+import * as web from '@opentelemetry/web';
+import { AttributeNames } from './enums/AttributeNames';
+import { FetchError, FetchResponse, SpanData } from './types';
+import { VERSION } from './version';
+
+// how long to wait for observer to collect information about resources
+// this is needed as event "load" is called before observer
+// hard to say how long it should really wait, seems like 300ms is
+// safe enough
+const OBSERVER_WAIT_TIME_MS = 300;
+
+/**
+ * FetchPlugin Config
+ */
+export interface FetchPluginConfig extends api.PluginConfig {
+  // the number of timing resources is limited, after the limit
+  // (chrome 250, safari 150) the information is not collected anymore
+  // the only way to prevent that is to regularly clean the resources
+  // whenever it is possible, this is needed only when PerformanceObserver
+  // is not available
+  clearTimingResources?: boolean;
+  // urls which should include trace headers when origin doesn't match
+  propagateTraceHeaderCorsUrls?: web.PropagateTraceHeaderCorsUrls;
+}
+
+/**
+ * This class represents a fetch plugin for auto instrumentation
+ */
+export class FetchPlugin extends core.BasePlugin<Promise<Response>> {
+  moduleName = 'fetch';
+  private _usedResources = new WeakSet<PerformanceResourceTiming>();
+  private _tasksCount = 0;
+
+  constructor(protected _config: FetchPluginConfig = {}) {
+    super('@opentelemetry/plugin-fetch', VERSION);
+  }
+
+  /**
+   * Add cors pre flight child span
+   * @param span
+   * @param corsPreFlightRequest
+   */
+  private _addChildSpan(
+    span: api.Span,
+    corsPreFlightRequest: PerformanceResourceTiming
+  ): void {
+    this._tracer.withSpan(span, () => {
+      const childSpan = this._tracer.startSpan('CORS Preflight', {
+        startTime: corsPreFlightRequest[web.PerformanceTimingNames.FETCH_START],
+      });
+      web.addSpanNetworkEvents(childSpan, corsPreFlightRequest);
+      childSpan.end(
+        corsPreFlightRequest[web.PerformanceTimingNames.RESPONSE_END]
+      );
+    });
```

Comment:
You can save yourself the overhead of calling the context manager here as this is always a "leaf" span (not calling any user code which may start another span or propagating).

```suggestion
    const childSpan = this._tracer.startSpan('CORS Preflight', {
      startTime: corsPreFlightRequest[web.PerformanceTimingNames.FETCH_START],
      parent: span
    });
    web.addSpanNetworkEvents(childSpan, corsPreFlightRequest);
    childSpan.end(
      corsPreFlightRequest[web.PerformanceTimingNames.RESPONSE_END]
    );
```